src/anyio/streams/memory.py

```diff
@@ -199,11 +206,14 @@ def send_nowait(self, item: T_contra) -> None:
         if not self._state.open_receive_channels:
             raise BrokenResourceError
 
-        if self._state.waiting_receivers:
-            receive_event, container = self._state.waiting_receivers.popitem(last=False)
-            container.append(item)
-            receive_event.set()
-        elif len(self._state.buffer) < self._state.max_buffer_size:
+        while self._state.waiting_receivers:
+            receive_event, receiver = self._state.waiting_receivers.popitem(last=False)
```

Comment:
I suspect that this can deadlock if the receiving task is cancelled, then send_nowait invoked (removing it from the waiting_receivers), and then _uncancelled_ before the event loop actually kicks the receiving task out of `await receive`.

Of course this is a pretty nasty use of the already brittle uncancel semantics, but we probably want a comment here to make the possibility salient to future readers?  Or maybe I'm wrong about this.

Reply:
I was not able to create a situation where an uncancel task would get its foot in the door to prevent the delivery of the cancellation. Here's my best attempt so far:
```py
import asyncio
from asyncio import CancelledError, current_task, get_running_loop

from anyio import create_memory_object_stream, create_task_group


async def receiver(receive, task_status):
    with receive:
        task_status.started(current_task())
        try:
            item = await receive.receive()
        except CancelledError:
            print("cancelled")
            raise

        print("got", item)


async def uncanceller(task):
    task.uncancel()
    print("uncancelled")


async def main():
    send, receive = create_memory_object_stream()
    async with create_task_group() as tg:
        task = await tg.start(receiver, receive)
        get_running_loop().call_soon(task.uncancel)
        task.cancel()
        await send.send(6)


asyncio.run(main())
```

Reply:
Looks like I was trying hard enough. I was able to repro the situation using just asyncio's task APIs:
```python
import asyncio
from asyncio import CancelledError, create_task

from anyio import create_memory_object_stream


async def receiver(receive):
    with receive:
        try:
            item = await receive.receive()
        except CancelledError:
            print("cancelled")
            raise

        print("got", item)


async def uncanceller(task):
    task.uncancel()
    print("uncancelled, cancelling =", task.cancelling())


async def main():
    send, receive = create_memory_object_stream()
    task = create_task(receiver(receive))
    uncancel_task = create_task(uncanceller(task))
    task.cancel()
    await send.send(6)


asyncio.run(main())
```

Reply:
Scratch that too, because the task hadn't even started and that's why the sender was blocking.

Reply:
Ok, here's my third attempt which seems to confirm that the cancellation is scheduled ahead of everything else:
```py
import asyncio
from asyncio import CancelledError, create_task, sleep

from anyio import create_memory_object_stream


async def receiver(receive):
    with receive:
        print("waiting for item")
        try:
            item = await receive.receive()
        except CancelledError:
            print("cancelled")
        else:
            print("got", item)


async def uncanceller(task):
    task.uncancel()
    print("uncancelled, cancelling =", task.cancelling())


async def main():
    send, receive = create_memory_object_stream(1)
    receive_task = create_task(receiver(receive))
    await sleep(0)
    uncancel_task = create_task(uncanceller(receive_task))
    receive_task.cancel()
    await send.send(6)
    await uncancel_task
    await receive_task


asyncio.run(main())
```
Output:
```
waiting for item
cancelled
uncancelled, cancelling = 0
Traceback (most recent call last):
...
```
Interestingly, if I cancel the receiver task and uncancel it right after without yielding to another task, I get different results based on the Python version. On 3.13.0b1, the task is scheduled normally and receives the item:
```py
import asyncio
from asyncio import CancelledError, create_task, sleep

from anyio import create_memory_object_stream


async def receiver(receive):
    with receive:
        print("waiting for item")
        try:
            item = await receive.receive()
        except CancelledError:
            print("cancelled")
        else:
            print("got", item)


async def main():
    send, receive = create_memory_object_stream()
    receive_task = create_task(receiver(receive))
    await sleep(0)
    receive_task.cancel()
    receive_task.uncancel()
    await send.send(6)
    await receive_task


asyncio.run(main())
```

Reply:
I suppose this is the effect of Guido's recent cancellation PR.